### PR TITLE
drivers: pwm: Add shim for nrfx PWM HW driver

### PR DESCRIPTION
**Pull request by jdemar**
Adds a translation layer to make the nrfx driver for nrf52 PWM
peripheral accessible via the Zephyr's API.

Signed-off-by: Justin DeMartino <jdemarti@gmail.com>

### CODE REVIEW

File: ext/hal/nordic/nrfx_config_nrf52840.h [hunk 1]
```diff
@@ -1826,16 +1826,16 @@
...
- 
-
-// <i> See more in the Errata document located at 
-// <i> https://infocenter.nordicsemi.com/
-
-#ifndef NRFX_SPIM3_NRF52840_ANOMALY_198_WORKAROUND_ENABLED
-#define NRFX_SPIM3_NRF52840_ANOMALY_198_WORKAROUND_ENABLED 0
-#endif
-
+    // <q> NRFX_SPIM3_NRF52840_ANOMALY_198_WORKAROUND_ENABLED  - Enables nRF52840 anomaly 198 workaround for SPIM3.
```
Comment (cpdesign) [inline ext/hal/nordic/nrfx_config_nrf52840.h]: did you mean to add this whitespace?

If so, I'd put it in a separate commit

File: ext/hal/nordic/Kconfig [hunk 2]
```diff
@@ -18,3 +18,6 @@ config NRFX_SPIM
 
 config NRFX_SPIS
 	bool
+
+config NRFX_PWM
+	bool
```
Comment (cpdesign) [inline ext/hal/nordic/Kconfig]: there is no newline at end of file

File: ext/hal/nordic/nrfx_config_nrf52832.h [hunk 3]
```diff
@@ -954,28 +954,28 @@
 
 // <e> NRFX_PWM_ENABLED - nrfx_pwm - PWM peripheral driver
 //==========================================================
-#ifndef NRFX_PWM_ENABLED
-#define NRFX_PWM_ENABLED 0
+#ifdef CONFIG_NRFX_PWM
+#define NRFX_PWM_ENABLED 1
```
Comment (cpdesign) [inline ext/hal/nordic/nrfx_config_nrf52832.h]: does this need another else clause to make sure that NRFX_PWM_ENABLED is set to 0 if it is not already defined?

eg:
#ifndef CONFIG_NRFX_PWM
#define NRFX_PWM_ENABLED 1
#else 
#ifndef NRFX_PWM_ENABLED
#define NRFX_PWM_ENABLED 0
#endif 
#endif

File: drivers/pwm/pwm_nrfx.c [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	nrfx_pwm_config_t config;
+};
+
+struct pwm_data {
+	nrf_pwm_sequence_t seq;
+	u16_t current[NRF_PWM_CHANNEL_COUNT];
+	u16_t top_value;
+};
+
+static int pwm_nrfx_pin_set(struct device *dev, u32_t pwm, u32_t
```
Comment (cpdesign) [inline drivers/pwm/pwm_nrfx.c]: Move the u32_t to the next line

File: drivers/pwm/Kconfig.nrfx [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+	bool "nRF HW PWM 0"
+	depends on SOC_SERIES_NRF52X
+	select NRFX_PWM
+	help
+	  Enable nRF Hardware PWM on port 0.
+
+if PWM_0_NRFX_PWM
+
+config PWM_0_NRFX_DEV_NAME
+	string "Nordic Semiconductor nRF52 series H/W PWM port  Device Name"
```
Comment (cpdesign) [inline drivers/pwm/Kconfig.nrfx]: I'd use 'module' instead of 'port' here (and similar other places)

File: drivers/pwm/Kconfig.nrfx [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+	help
+		Clock = 16MHz / (2^prescaler)
+
+config PWM_0_NRFX_CH0_PIN
+	int "CH0 pin number"
+	range -1 47 if SOC_NRF52840_QIAA
+	range -1 31
+	default -1
+	help
+	  GPIO pin number to use for CH0
```
Comment (cpdesign) [inline drivers/pwm/Kconfig.nrfx]: On first read I wondered why the range starts from -1; guessing that is the value for 'not used'

Perhaps an extra "(set to -1 if not used)" in the comment would be nice here.

Alternatively, could you have an enable option for each pin, and then another option to set the range.

File: ext/hal/nordic/nrfx_config_nrf52832.h (hunk 3 above)
Comment (jdemar) [reply ext/hal/nordic/nrfx_config_nrf52832.h]: I followed the NRFX_SPIM driver as my example.  I dont think its needed based on how it implements the NRFX_CHECK macro.

File: ext/hal/nordic/nrfx_config_nrf52840.h (hunk 1 above)
Comment (jdemar) [reply ext/hal/nordic/nrfx_config_nrf52840.h]: I'll fix it

File: drivers/pwm/pwm_nrfx.c [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+			},													      \
+			.irq_priority = NRFX_PWM_DEFAULT_CONFIG_IRQ_PRIORITY,							      \
+			.base_clock = (nrf_pwm_clk_t)CONFIG_PWM_##idx##_NRFX_CLOCK_PRESCALER,					      \
+			.count_mode = NRF_PWM_MODE_UP,										      \
+			.top_value = NRFX_PWM_DEFAULT_CONFIG_TOP_VALUE,								      \
+			.load_mode = NRF_PWM_LOAD_INDIVIDUAL,									      \
+			.step_mode = NRF_PWM_STEP_TRIGGERED,									      \
+		}														      \
+	};															      \
+	static struct pwm_data pwm_nrfx_##idx##_data = {									      \
```
Comment (anangl) [inline drivers/pwm/pwm_nrfx.c]: Adding `NRFX_PWM_PIN_INVERTED` to a pin number is not enough to make a pin inverted, since it only defines GPIO configuration of the pin. For an inverted channel you need to set bit 15 in the value supplied for this channel (see [doc](https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/nordic/nrfx/hal/nrf_pwm.h#L244)).
Therefore, I propose to initialize the channels here with 0 for normal and 0x8000 for inverted ones, and keep bit 15 untouched while setting a new value for a channel in `pwm_nrfx_pin_set`.

File: drivers/pwm/pwm_nrfx.c [hunk 8]
```diff
@@ -0,0 +1,202 @@
+#include <nrfx_pwm.h>
+#include "pwm.h"
+
+#define SYS_LOG_DOMAIN "pwm/nrf5x"
```
Comment (anangl) [inline drivers/pwm/pwm_nrfx.c]: "pwm_nrfx"

**Review by anangl — CHANGES_REQUESTED** [review]

File: drivers/pwm/pwm_nrfx.c (hunk 7 above)
Comment (jdemar) [reply drivers/pwm/pwm_nrfx.c]: I'll make the change, but from my testing, it seems like if bit 15 is set its normal, not set is inverted.  It's not really clear in that doc.

Comment (anangl) [reply drivers/pwm/pwm_nrfx.c]: You're right. I should be as you wrote - bit 15 set for normal (i.e. pin at 0 when idle, switching to 1 at compare match) channels and not set for inverted ones.
My bad, sorry.

**Comment by jdemar** [comment]
@anangl I've made the requested changes

File: drivers/pwm/pwm_nrfx.c [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+	u16_t top_value;
+};
+
+static int pwm_nrfx_pin_set(struct device *dev, u32_t pwm,
+			    u32_t period_cycles, u32_t pulse_cycles)
+{
+	const struct pwm_nrfx_config *pconfig = dev->config->config_info;
+	struct pwm_nrfx_data *pdata = dev->driver_data;
+
+	if (pwm > NRF_PWM_CHANNEL_COUNT) {
```
Comment (anangl) [inline drivers/pwm/pwm_nrfx.c]: `>=`

File: drivers/pwm/pwm_nrfx.c [hunk 10]
```diff
@@ -0,0 +1,216 @@
+#include <nrfx_pwm.h>
```
Comment (anangl) [inline drivers/pwm/pwm_nrfx.c]: Add a license header.

File: drivers/pwm/pwm_nrfx.c [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+	static const struct pwm_nrfx_config pwm_nrfx_##idx##_config = {			     \
+	.pwm = NRFX_PWM_INSTANCE(idx),							     \
+	.config = {									     \
+		.output_pins = {							     \
+			PWM_NRFX_OUTPUT_PIN(idx, 0),					     \
+			PWM_NRFX_OUTPUT_PIN(idx, 1),					     \
+			PWM_NRFX_OUTPUT_PIN(idx, 2),					     \
+			PWM_NRFX_OUTPUT_PIN(idx, 3),					     \
+		},									     \
+		.irq_priority = NRFX_PWM_DEFAULT_CONFIG_IRQ_PRIORITY,			     \
```
Comment (anangl) [inline drivers/pwm/pwm_nrfx.c]: This can be skipped, to not create confusion - interrupts are not used by this driver, and even if they were, the priority would need to be set via IRQ_CONNECT (this field is simply ignored).

File: drivers/pwm/pwm_nrfx.c [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+	}
+
+	if (period_cycles != pdata->top_value) {
+		if (period_cycles > PWM_COUNTERTOP_COUNTERTOP_Msk) {
+			return -EINVAL;
+		}
+
+		pdata->top_value = period_cycles;
+		nrf_pwm_configure(pconfig->pwm.p_registers,
+				  pconfig->config.base_clock, pconfig->config.count_mode,
```
Comment (anangl) [inline drivers/pwm/pwm_nrfx.c]: Place each parameter in a separate line.

**Review by anangl — CHANGES_REQUESTED** [review]
Some lines exceed the 80 characters limit. I am a bit surprised that CI does not complain about that. But this should be corrected.

File: drivers/pwm/Kconfig.nrfx [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+if PWM_NRFX
+
+config PWM_0_NRFX_PWM
+	bool "nRF HW PWM 0"
+	depends on SOC_SERIES_NRF52X
+	select NRFX_PWM
+	help
+	  Enable nRF Hardware PWM module 0.
+
+if PWM_0_NRFX_PWM
```
[the remark this answers is not in the document]
Comment (jdemar) [reply drivers/pwm/Kconfig.nrfx]: Should PWM_0_NRFX_DEV_NAME then be PWM_0_DEV_NAME?

[the remark this answers is not in the document]
Comment (anangl) [reply drivers/pwm/Kconfig.nrfx]: Actually, `PWM_0_NAME`. To keep consistency with what we've got for SPI, I2C, or ADC.

And for other config entries here, I think we should use the convention `PWM_x_NRF_...`, since they are not strictly related to the nrfx driver but rather to the nRF peripherals themselves. We do the same in SPI, I2C (in #7788), or UART.

File: ext/hal/nordic/Kconfig [hunk 14]
```diff
@@ -19,5 +19,8 @@ config NRFX_SPIM
 config NRFX_SPIS
 	bool
 
+config NRFX_PWM
```
Comment (anangl) [inline ext/hal/nordic/Kconfig]: Keep the alphabetical order here, please.

**Review by anangl — CHANGES_REQUESTED** [review]

**Review by anangl — CHANGES_REQUESTED** [review]

File: drivers/pwm/pwm_nrfx.c [hunk 15]
```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (jdemar) [reply drivers/pwm/pwm_nrfx.c]: Fixed

File: drivers/pwm/pwm_nrfx.c [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+				  pdata->top_value);
+	}
+
+	if (pulse_cycles > pdata->top_value) {
+		return -EINVAL;
+	}
+
+	pdata->current[pwm] = ((pdata->current[pwm]
+				& ~PWM_COUNTERTOP_COUNTERTOP_Msk)
+				| pulse_cycles);
```
[the remark this answers is not in the document]
Comment (jdemar) [reply drivers/pwm/pwm_nrfx.c]: Fixed

**Review by anangl — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
